packages/cli/src/controller/codegen-controller.ts

```diff
@@ -155,6 +167,106 @@ export async function generateEnums(projectPath: string, schema: string): Promis
   }
 }
 
+interface abiRenderProps {
+  name: string;
+  events: string[];
+  functions: {typeName: string; functionName: string}[];
+}
+interface abiInterface {
+  name: string;
+  type: 'event' | 'function';
+  inputs: {
+    internalType: string;
+    name: string;
+    type: string;
+  }[];
+}
+export async function generateAbis(datasources: DatasourceKind[], projectPath: string): Promise<void> {
+  const sortedAssets = new Map();
+  datasources.map((d) => {
+    Object.entries(d.assets).map(([name, value]) => {
+      const filePath = path.join(projectPath, value.file);
+      if (!fs.existsSync(filePath)) {
+        throw new Error(`Error: Asset ${name}, file ${value.file} does not exist`);
+      }
+      // We use actual abi file name instead on name provided in assets
+      // This is aligning with files in './ethers-contracts'
+      sortedAssets.set(parseContractPath(filePath).name, value.file);
+    });
+  });
+  if (sortedAssets.size !== 0) {
+    await prepareDirPath(path.join(projectPath, ABI_INTERFACES_ROOT_DIR), true);
+    try {
+      const allFiles = glob(projectPath, [...sortedAssets.values()]);
+      await runTypeChain({
+        cwd: projectPath,
+        filesToProcess: allFiles,
+        allFiles,
+        outDir: ETHERS_CONTRACTS_DIR,
+        target: TYPECHAIN_TARGET,
+      });
+      // we iterate here as we have to make sure type chain generated successful,
+      // also avoid duplicate generate same abi interfaces
+      const renderInterfaceJobs: abiRenderProps[] = [];
```

Comment:
Could we move from here down into another function to break it up and be more readable?